### PR TITLE
Enable shared code between /web and /server

### PR DESCRIPTION
**Pull request by rileyjshaw**
Until now, we've been able to share types from /common to both our
/server and /web codebases. However, importing any code into /server
from /common that wasn't stripped out by TypeScript (so, anything but
types), caused our directory structure to change. It broke the build
in a non-obvious way, and has been a frustrating limitation.

I've slowly and surely been upgrading our Webpack / TypeScript
dependencies, which was part of the puzzle.

This commit turns our /common directory into a proper standalone module,
with a single generated index.ts file, a build step, and type
annotations.

For consistency, we now rely on [Yarn workspaces](https://classic.yarnpkg.com/en/docs/workspaces/) on both the /web and /server packages. Treating /common as a standalone module has encapsulation benefits, and will let us manage a separate dependency list, build steps, etc. for /common.

In the process of doing this work I realized that we have a lot of (near) options for solving this task. Here are a few off the top of my head:

- Yarn Workspaces
- Yarn / NPM Private Packages
- Yarn link
- "file: ../" dependencies
- TypeScript Project References
- TypeScript Internal Modules
- TypeScript Path Aliases
- TypeScript Path Mapping (different thing)
- Lerna
- Oh… have I not even mentioned webpack yet?

We already using some of these, which made the task frustrating. The goal is to pick a subset that works for us, and strictly stick to that subset.

This commit also adds some trivial `real code` imports into /server,
so we can test it out right away.

For quick checks, I found the following adequate:

`cd server && tsc -b && cd ../web && yarn build`

or,

`cd server && yarn start` (tests whether node / commonjs imports are
finding /common imports)

or,

`cd common && tsc`

To run a final test of this change, I wanted to be sure that things would work
from scratch. The test plan below is a bit overkill, and it will take a while,
but it was worth the peace of mind.

```
docker ps -a
docker images
rm -rf web/node_modules/ server/node_modules/ node_modules/ web/dist/ server/js common/js

docker-compose up
```

After waiting for a long time, I went to localhost:9000 and… it worked!  🎉

The current monorepo structure was initially introduced in
https://github.com/mozilla/voice-web/commit/edcf9ae56948b517ce505c6af7a975f0790b3db7

### CODE REVIEW

**Comment by rileyjshaw** [comment]
^ rebase

**Comment by rileyjshaw** [comment]
- [x] Fixme: Remove all @ts-ignore lines.
- [x] Todo: ensure bundle sizes haven't exploded.
- [x] Todo: Ensure nothing has gotten in the way of our IDE-ish TypeScript capabilities.

All done! Bundle size is unaffected, and TypeScript is still properly checking across packages. Revised the PR message to remove this list.

**Comment by rileyjshaw** [comment]
😓 apologies for the noise on this PR! Jest is having trouble finding the common module, and since our test config errors locally, I don't know if I've fixed it until it hits Travis.

We should totally fix local testing.

**Comment by phirework** [comment]
Love that this is that extremely rare PR where 52 files changed doesn't actually result in a lot for me to look at 😅 I've tested this locally and here are the things I've noticed, which I don't think is related to your config changes and more just the general fragility of our deploy setup: 

* The first time I ran the app after deleting everything and starting totally from scratch, the frontend failed, which fixed itself after I killed and restarted normally (ctrl+c):
```web        | [FE] [BABEL] Note: The code generator has deoptimised the styling of /code/node_modules/react-dom/cjs/react-dom.development.js as it exceeds the max of 500KB.
web        | [FE] [BABEL] Note: The code generator has deoptimised the styling of /code/node_modules/lottie-web/build/player/lottie.js as it exceeds the max of 500KB.
web        | [FE] Killed
web        | [FE] error Command failed with exit code 137.
web        | [FE] info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
web        | [FE] cd web && yarn dev exited with code 137```

* It couldn't find the S3proxy bucket - the startup code is showing: 
```web        | GET bucket code: 000
web        | Bucket does not exist. Creating one...
web        | PUT bucket code: 000
web        | Done!```

and attempting to record a clip results in a bunch of these:
```web        | [BE] { NoSuchBucket: The specified bucket does not exist
web        | [BE]     at Request.extractError (/code/node_modules/aws-sdk/lib/services/s3.js:816:35)
web        | [BE]     at Request.callListeners (/code/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
web        | [BE]     at Request.emit (/code/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
web        | [BE]     at Request.emit (/code/node_modules/aws-sdk/lib/request.js:683:14)
web        | [BE]     at Request.transition (/code/node_modules/aws-sdk/lib/request.js:22:10)
web        | [BE]     at AcceptorStateMachine.runTo (/code/node_modules/aws-sdk/lib/state_machine.js:14:12)
web        | [BE]     at /code/node_modules/aws-sdk/lib/state_machine.js:26:10
web        | [BE]     at Request.<anonymous> (/code/node_modules/aws-sdk/lib/request.js:38:9)
web        | [BE]     at Request.<anonymous> (/code/node_modules/aws-sdk/lib/request.js:685:12)
web        | [BE]     at Request.callListeners (/code/node_modules/aws-sdk/lib/sequential_executor.js:116:18)
web        | [BE]     at Request.emit (/code/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
web        | [BE]     at Request.emit (/code/node_modules/aws-sdk/lib/request.js:683:14)
web        | [BE]     at Request.transition (/code/node_modules/aws-sdk/lib/request.js:22:10)
web        | [BE]     at AcceptorStateMachine.runTo (/code/node_modules/aws-sdk/lib/state_machine.js:14:12)
web        | [BE]     at /code/node_modules/aws-sdk/lib/state_machine.js:26:10
web        | [BE]     at Request.<anonymous> (/code/node_modules/aws-sdk/lib/request.js:38:9)
web        | [BE]   message: 'The specified bucket does not exist',
web        | [BE]   code: 'NoSuchBucket',
web        | [BE]   region: null,
web        | [BE]   time: 2020-03-02T21:58:48.888Z,
web        | [BE]   requestId: '4442587FB7D0A2F9',
web        | [BE]   extendedRequestId: undefined,
web        | [BE]   cfId: undefined,
web        | [BE]   statusCode: 404,
web        | [BE]   retryable: false,
web        | [BE]   retryDelay: 96.29193391756385 }```

I ran the `prepare_s3.sh` manually from inside the Docker container and it was fine (the returned codes were 404 and 200, respectively). I suspect this is just something wonky with my Docker setup, but wanted to check to see if you had the same experience?

I've clicked around and tested login and everything seems to be fine there! I would hold off on merging this into master until after the release tomorrow, we'll want to give this a couple of weeks on staging to properly test before sending it out into the wild.

File: web/src/hooks/use-active-section.tsx [hunk 1]
```diff
@@ -1,76 +0,0 @@
...
-
-    sectionsElements.map((e: HTMLElement) => observer.observe(e));
-
-    return () => {
-      sectionsElements.map((e: HTMLElement) => observer.unobserve(e));
-    };
-  });
-
-  return activeSection;
-}
```
[the remark this answers is not in the document]
Comment (rileyjshaw) [reply web/src/hooks/use-active-section.tsx]: Ah, I should have left a comment about this. I'm not sure when this file was added, but there is a duplicate, non-tsx file (`web/src/hooks/use-active-section.ts`) that's actually being included wherever we `import … from 'use-active-section'`.

**Comment by phirework** [comment]
Hurray for tests passing! Do you want to try deploying by tagging this branch first before merging into master? Feel free to skip if you feel confident about not needing to revert anything before the next release but I thought it might be a low-stakes way of testing in our env.

**Comment by rileyjshaw** [comment]
@phirework heck yesssss, this is the perfect opportunity to test out a staging environment 😍 great idea.

I'm still looking into your initial feedback by the way – I have some tests running in the background and will get back with a proper response once they're done.

**Comment by rileyjshaw** [comment]
@phirework re: [your comments above](https://github.com/mozilla/voice-web/pull/2594#issuecomment-593647802):

1. On a completely fresh environment (new docker container, completely wiped build files and modules), my frontend built without error. After a long wait, my webserver started and served our site as expected!

```
web        | [FE] [BABEL] Note: The code generator has deoptimised the styling of /code/node_modules/lodash/lodash.js as it exceeds the max of 500KB.
web        | [FE] [BABEL] Note: The code generator has deoptimised the styling of /code/node_modules/react-dom/cjs/react-dom.development.js as it exceeds the max of 500KB.
web        | [FE] [BABEL] Note: The code generator has deoptimised the styling of /code/node_modules/lottie-web/build/player/lottie.js as it exceeds the max of 500KB.
web        | [FE] Built!
```

2. I also got this output!

```
web        | GET bucket code: 000
web        | Bucket does not exist. Creating one...
db         | 2020-03-05 20:25:33 1 [Note] Event Scheduler: Loaded 0 events
db         | 2020-03-05 20:25:33 1 [Note] mysqld: ready for connections.
db         | Version: '5.6.39'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server (GPL)
web        | PUT bucket code: 000
web        | Done!
```

I see that on master as well, and assumed it had something to do with my brand new Docker container. Let me know if that's concerning to you!

3. The /speak page loads fine for me, but afteri recording a few clips and hitting submit I see similar issues. I get the same issue on `master`, though, so I don't think it's specific to this PR?

---

I pushed the tag `staging-v1.23.0-codeshare` a few hours ago to test the changes on our staging environment. Sadly, the /speak page didn't even fully load 😞 I just pushed a new tag, `staging-v1.23.0-master-test-1`, to see if this is a staging server issue or a bug in the PR.

**Pull request merged**